gqcp/include/QCModel/HF/GHF.hpp

```diff
@@ -0,0 +1,395 @@
+
+/**
+ *  The generalized Hartree-Fock wave function model.
+ * 
+ *  @tparam _Scalar             the type of scalar that is used for the expansion of the spatial orbitals in their underlying scalar basis
+ */
+template <typename _Scalar>
+class GHF {
+public:
+    using Scalar = _Scalar;
+
+
+private:
+    size_t N;  // the number of electrons
+
+    VectorX<double> orbital_energies;  // sorted in ascending energies
+    TransformationMatrix<Scalar> C;    // the coefficient matrix that expresses every spinor orbital (as a column) in the underlying scalar bases
+
+
+public:
+    /*
+     *  CONSTRUCTORS
+     */
+
+    /**
+     *  The standard member-wise constructor
+     * 
+     *  @param N                    the number of electrons
+     *  @param C                    the coefficient matrix that expresses every spinor orbital (as a column) in the underlying scalar bases
+     *  @param orbital_energies     the GHF MO energies
+     */
+    GHF(const size_t N, const VectorX<double>& orbital_energies, const TransformationMatrix<double>& C) :
+        N {N},
+        orbital_energies {orbital_energies},
+        C(C) {}
+
+
+    /**
+     *  Default constructor setting everything to zero
+     */
+    GHF() :
+        GHF(0, TransformationMatrix<double>::Zero(0, 0), VectorX<double>::Zero(0)) {}
+
+
+    /*
+     *  STATIC PUBLIC METHODS
+     */
+
+    /**
+     *  @param F                the Fock matrix expressed in the scalar bases
+     *  @param D                the GHF density matrix in the same scalar bases
+     *  @param S                the (spin-blocked) overlap matrix of the scalar bases
+     * 
+     *  @return the GHF error matrix
+     */
+    static SquareMatrix<Scalar> calculateError(const QCMatrix<Scalar>& F, const OneDM<Scalar>& P, const SquareMatrix<Scalar>& S) {
+        return F * P * S - S * P * F;
+    }
+
+
+    /**
+     *  @param P                the (spin-blocked) GHF density matrix in the scalar bases
+     *  @param H_core           the (spin-blocked) core Hamiltonian expressed in the same scalar bases
+     *  @param F                the (spin-blocked) Fock matrix in the same scalar bases
+     *
+     *  @return the GHF electronic energy
+     */
+    static double calculateElectronicEnergy(const OneDM<Scalar>& P, const ScalarSQOneElectronOperator<Scalar>& H_core, const ScalarSQOneElectronOperator<Scalar>& F) {
+
+        // First, calculate the sum of H_core and F (this saves a contraction)
+        ScalarSQOneElectronOperator<Scalar> Z = H_core + F;
+
+        // Convert the matrices Z and D to an Eigen::Tensor<double, 2> D_tensor, as contractions are only implemented for Tensors
+        Eigen::TensorMap<Eigen::Tensor<const Scalar, 2>> P_tensor {P.data(), P.rows(), P.cols()};
+        Eigen::TensorMap<Eigen::Tensor<double, 2>> Z_tensor {Z.parameters().data(), P.rows(), P.cols()};
+
+        // Specify the contraction pair.
+        // To calculate the electronic energy, we must perform a double contraction.
```

Comment:
Since `knowdes` is publicly available now, maybe we could add a reference to the formula?

Reply:
Do you mean a link to `knowdes' Generalized Hartree-Fock` page? Or a reference to Stuber & Paldus in which the energy formula is derived?